### PR TITLE
fix for #4029 apt-get autoremove

### LINKED ISSUE #4029: Error with module apt autoremove
##### ISSUE TYPE
- Bug Report
##### COMPONENT NAME

apt autoremove
##### ANSIBLE VERSION

``` bash
ansible 2.1.0.0
  config file = /home/galindro/git_repos/devops/ansible.cfg
  configured module search path = Default w/o overrides
```
##### CONFIGURATION

``` bash
$ cat ansible.cfg

[defaults]
host_key_checking = False

[ssh_connection]
pipelining=False
ssh_args = -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o ControlPath="~/.ansible/tmp/ansible-ssh-%h-%p-%r"
```

I'm NOT using any ANSIBLE_\* environment variable
##### OS / ENVIRONMENT

```
No LSB modules are available.
Distributor ID: Debian
Description:    Debian GNU/Linux 8.5 (jessie)
Release:    8.5
Codename:   jessie
```
##### SUMMARY

I'm used the apt module with autoremove option and this error was showed when I run the task:

```
TASK [common : autoremove unused packages] *************************************
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```
##### STEPS TO REPRODUCE

Create this task:

``` yaml
- name: autoremove unused packages
  apt:
    autoremove: yes
  sudo: yes
  tags:
    - infra
```

Put that task inside a simple role

Create a playbook:

``` yaml
- hosts: all
  role:
     - myrolename
  environment:
    TMOUT: 600
```

Run the playbook

```
ansible-playbook -i ec2.py -v -b myplay.yml
```

P.S.: my host is an instance running on amazon ec2
##### EXPECTED RESULTS

This error:

``` bash
TASK [myrolename : autoremove unused packages] *************************************
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_RpjDPt/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```
##### ACTUAL RESULTS

Running with -vvvv:

```
TASK [myrolename : autoremove unused packages] *************************************
task path: /home/galindro/git_repos/devops/roles/myrolename/tasks/debian.yml:52
<172.30.2.214> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.30.2.214> SSH: EXEC ssh -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 172.30.2.214 '/bin/sh -c '"'"'( umask 77 && mkdir -p "` echo $HOME/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254 `" && echo ansible-tmp-1466722720.22-229804017504254="` echo $HOME/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254 `" ) && sleep 0'"'"''
<172.30.2.214> PUT /tmp/tmpBiapsW TO /home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/apt
<172.30.2.214> SSH: EXEC sftp -b - -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 '[172.30.2.214]'
<172.30.2.214> ESTABLISH SSH CONNECTION FOR USER: ubuntu
<172.30.2.214> SSH: EXEC ssh -C -vvv -o ForwardAgent=yes -o ControlMaster=auto -o ControlPersist=30m -o 'ControlPath=~/.ansible/tmp/ansible-ssh-%h-%p-%r' -o StrictHostKeyChecking=no -o Port=40 -o 'IdentityFile="sbkey_20150204_app.pem"' -o KbdInteractiveAuthentication=no -o PreferredAuthentications=gssapi-with-mic,gssapi-keyex,hostbased,publickey -o PasswordAuthentication=no -o User=ubuntu -o ConnectTimeout=10 -tt 172.30.2.214 '/bin/sh -c '"'"'sudo -H -S -n -u root /bin/sh -c '"'"'"'"'"'"'"'"'echo BECOME-SUCCESS-yhgovfdopoujnzwegznjpbbaoqcpcnyd; LANG=en_US.UTF-8 LC_ALL=en_US.UTF-8 TMOUT=600 LC_MESSAGES=en_US.UTF-8 /usr/bin/python /home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/apt; rm -rf "/home/ubuntu/.ansible/tmp/ansible-tmp-1466722720.22-229804017504254/" > /dev/null 2>&1'"'"'"'"'"'"'"'"' && sleep 0'"'"''
fatal: [172.30.2.214]: FAILED! => {"changed": false, "failed": true, "invocation": {"module_name": "apt"}, "module_stderr": "OpenSSH_6.7p1 Debian-5+deb8u2, OpenSSL 1.0.1t  3 May 2016\r\ndebug1: Reading configuration data /home/galindro/.ssh/config\r\ndebug1: Reading configuration data /etc/ssh/ssh_config\r\ndebug1: /etc/ssh/ssh_config line 19: Applying options for *\r\ndebug1: auto-mux: Trying existing master\r\ndebug2: fd 3 setting O_NONBLOCK\r\ndebug2: mux_client_hello_exchange: master version 4\r\ndebug3: mux_client_forwards: request forwardings: 0 local, 0 remote\r\ndebug3: mux_client_request_session: entering\r\ndebug3: mux_client_request_alive: entering\r\ndebug3: mux_client_request_alive: done pid = 32546\r\ndebug3: mux_client_request_session: session request sent\r\ndebug1: mux_client_request_session: master session id: 2\r\ndebug3: mux_client_read_packet: read header failed: Broken pipe\r\ndebug2: Received exit status from master 0\r\nShared connection to 172.30.2.214 closed.\r\n", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_uvVsjU/ansible_module_apt.py\", line 842, in <module>\r\n    main()\r\n  File \"/tmp/ansible_uvVsjU/ansible_module_apt.py\", line 802, in main\r\n    for package in packages:\r\nTypeError: 'NoneType' object is not iterable\r\n", "msg": "MODULE FAILURE", "parsed": false}
```

### PR DESCRIPTION
**Pull request by gdelpierre**
##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
 ```
packaging/os/apt.py
```

##### ANSIBLE VERSION

```
2.3
```

##### SUMMARY

Fixes #4029 

###### ALL UNUSED PACKAGES
```
- name: Apt-repositories | Autoremove unused packages
  apt:
    autoremove: 'yes'
```

###### RESULT
```
changed: [localhost] => {
    "changed": true,
    "diff": {},
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoremove": true,
            "cache_valid_time": 0,
            "deb": null,
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "only_upgrade": false,
            "package": null,
            "purge": false,
            "state": "present",
            "update_cache": false,
            "upgrade": null
        },
        "module_name": "apt"
    },
    "stderr": "invoke-rc.d: policy-rc.d denied execution of stop.\n",
    "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages will be REMOVED:\n  fakeroot krb5-locales libalgorithm-diff-perl libalgorithm-diff-xs-perl\n  libalgorithm-merge-perl libfakeroot libfile-fcntllock-perl libglib2.0-data\n  libjs-jquery libsasl2-modules libssl-doc libx11-6 libx11-data libxau6\n  libxcb1 libxdmcp6 libxext6 libxml2 libxmuu1 manpages manpages-dev\n  ncurses-term python-requests python-urllib3 rsync sgml-base shared-mime-info\n  ssh-import-id tcpd wget xauth xml-core\n0 upgraded, 0 newly installed, 32 to remove and 0 not upgraded.\nAfter this operation, 22.4 MB disk space will be freed.\n(Reading database ... 28316 files and directories currently installed.)\nRemoving fakeroot (1.20-3ubuntu2) ...\nupdate-alternatives: using /usr/bin/fakeroot-tcp to provide /usr/bin/fakeroot (fakeroot) in auto mode\nRemoving krb5-locales (1.12+dfsg-2ubuntu5.2) ...\nRemoving libalgorithm-merge-perl (0.08-2) ...\nRemoving libalgorithm-diff-xs-perl (0.04-2build4) ...\nRemoving libalgorithm-diff-perl (1.19.02-3) ...\nRemoving libfakeroot:amd64 (1.20-3ubuntu2) ...\nRemoving libfile-fcntllock-perl (0.14-2build1) ...\nRemoving libglib2.0-data (2.40.2-0ubuntu1) ...\nRemoving libjs-jquery (1.7.2+dfsg-2ubuntu1) ...\nRemoving libsasl2-modules:amd64 (2.1.25.dfsg1-17build1) ...\nRemoving libssl-doc (1.0.1f-1ubuntu2.21) ...\nRemoving xauth (1:1.0.7-1ubuntu1) ...\nRemoving libxext6:amd64 (2:1.3.2-1ubuntu0.0.14.04.1) ...\nRemoving shared-mime-info (1.2-0ubuntu3) ...\nRemoving libxml2:amd64 (2.9.1+dfsg1-3ubuntu4.8) ...\nRemoving libxmuu1:amd64 (2:1.1.1-1) ...\nRemoving manpages-dev (3.54-1ubuntu1) ...\nRemoving manpages (3.54-1ubuntu1) ...\nRemoving ncurses-term (5.9+20140118-1ubuntu1) ...\nRemoving ssh-import-id (3.21-0ubuntu1) ...\nRemoving python-requests (2.2.1-1ubuntu0.3) ...\nRemoving python-urllib3 (1.7.1-1ubuntu4) ...\nRemoving rsync (3.1.0-2ubuntu0.2) ...\nRemoving xml-core (0.13+nmu2) ...\nRemoving sgml-base (1.26+nmu4ubuntu1) ...\nRemoving tcpd (7.6.q-25) ...\nRemoving wget (1.15-1ubuntu1.14.04.2) ...\nRemoving libx11-6:amd64 (2:1.6.2-1ubuntu2) ...\nRemoving libx11-data (2:1.6.2-1ubuntu2) ...\nRemoving libxcb1:amd64 (1.10-2ubuntu1) ...\nRemoving libxau6:amd64 (1:1.0.8-1) ...\nRemoving libxdmcp6:amd64 (1:1.1.1-1) ...\nProcessing triggers for libc-bin (2.19-0ubuntu6.9) ...\n",
    "stdout_lines": [
        "Reading package lists...",
        "Building dependency tree...",
        "Reading state information...",
        "The following packages will be REMOVED:",
        "  fakeroot krb5-locales libalgorithm-diff-perl libalgorithm-diff-xs-perl",
        "  libalgorithm-merge-perl libfakeroot libfile-fcntllock-perl libglib2.0-data",
        "  libjs-jquery libsasl2-modules libssl-doc libx11-6 libx11-data libxau6",
        "  libxcb1 libxdmcp6 libxext6 libxml2 libxmuu1 manpages manpages-dev",
        "  ncurses-term python-requests python-urllib3 rsync sgml-base shared-mime-info",
        "  ssh-import-id tcpd wget xauth xml-core",
        "0 upgraded, 0 newly installed, 32 to remove and 0 not upgraded.",
        "After this operation, 22.4 MB disk space will be freed.",
        "(Reading database ... 28316 files and directories currently installed.)",
        "Removing fakeroot (1.20-3ubuntu2) ...",
        "update-alternatives: using /usr/bin/fakeroot-tcp to provide /usr/bin/fakeroot (fakeroot) in auto mode",
        "Removing krb5-locales (1.12+dfsg-2ubuntu5.2) ...",
        "Removing libalgorithm-merge-perl (0.08-2) ...",
        "Removing libalgorithm-diff-xs-perl (0.04-2build4) ...",
        "Removing libalgorithm-diff-perl (1.19.02-3) ...",
        "Removing libfakeroot:amd64 (1.20-3ubuntu2) ...",
        "Removing libfile-fcntllock-perl (0.14-2build1) ...",
        "Removing libglib2.0-data (2.40.2-0ubuntu1) ...",
        "Removing libjs-jquery (1.7.2+dfsg-2ubuntu1) ...",
        "Removing libsasl2-modules:amd64 (2.1.25.dfsg1-17build1) ...",
        "Removing libssl-doc (1.0.1f-1ubuntu2.21) ...",
        "Removing xauth (1:1.0.7-1ubuntu1) ...",
        "Removing libxext6:amd64 (2:1.3.2-1ubuntu0.0.14.04.1) ...",
        "Removing shared-mime-info (1.2-0ubuntu3) ...",
        "Removing libxml2:amd64 (2.9.1+dfsg1-3ubuntu4.8) ...",
        "Removing libxmuu1:amd64 (2:1.1.1-1) ...",
        "Removing manpages-dev (3.54-1ubuntu1) ...",
        "Removing manpages (3.54-1ubuntu1) ...",
        "Removing ncurses-term (5.9+20140118-1ubuntu1) ...",
        "Removing ssh-import-id (3.21-0ubuntu1) ...",
        "Removing python-requests (2.2.1-1ubuntu0.3) ...",
        "Removing python-urllib3 (1.7.1-1ubuntu4) ...",
        "Removing rsync (3.1.0-2ubuntu0.2) ...",
        "Removing xml-core (0.13+nmu2) ...",
        "Removing sgml-base (1.26+nmu4ubuntu1) ...",
        "Removing tcpd (7.6.q-25) ...",
        "Removing wget (1.15-1ubuntu1.14.04.2) ...",
        "Removing libx11-6:amd64 (2:1.6.2-1ubuntu2) ...",
        "Removing libx11-data (2:1.6.2-1ubuntu2) ...",
        "Removing libxcb1:amd64 (1.10-2ubuntu1) ...",
        "Removing libxau6:amd64 (1:1.0.8-1) ...",
        "Removing libxdmcp6:amd64 (1:1.1.1-1) ...",
        "Processing triggers for libc-bin (2.19-0ubuntu6.9) ..."
    ]
}
```

or if nothing to remove
```
ok: [localhost] => {
    "changed": false,
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoremove": true,
            "cache_valid_time": 0,
            "deb": null,
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "only_upgrade": false,
            "package": null,
            "purge": false,
            "state": "present",
            "update_cache": false,
            "upgrade": null
        },
        "module_name": "apt"
    },
    "stderr": "",
    "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\n0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.\n",
    "stdout_lines": [
        "Reading package lists...",
        "Building dependency tree...",
        "Reading state information...",
        "0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded."
    ]
}
```

###### SPECIFIC UNUSED PACKAGE
```
- name: Apt-repositories | Autoremove unused packages
  apt:
    name: 'man-db'
    autoremove: 'yes'
```

###### RESULT
```
changed: [localhost] => {
    "changed": true,
    "diff": {},
    "invocation": {
        "module_args": {
            "allow_unauthenticated": false,
            "autoremove": true,
            "cache_valid_time": 0,
            "deb": null,
            "default_release": null,
            "dpkg_options": "force-confdef,force-confold",
            "force": false,
            "install_recommends": null,
            "name": "man-db",
            "only_upgrade": false,
            "package": [
                "man-db"
            ],
            "purge": false,
            "state": "present",
            "update_cache": false,
            "upgrade": null
        },
        "module_name": "apt"
    },
    "stderr": "",
    "stdout": "Reading package lists...\nBuilding dependency tree...\nReading state information...\nThe following packages will be REMOVED:\n  bsdmainutils groff-base libpipeline1 man-db\n0 upgraded, 0 newly installed, 4 to remove and 0 not upgraded.\nAfter this operation, 5635 kB disk space will be freed.\n(Reading database ... 22252 files and directories currently installed.)\nRemoving man-db (2.6.7.1-1ubuntu1) ...\nRemoving bsdmainutils (9.0.5ubuntu1) ...\nRemoving groff-base (1.22.2-5) ...\nRemoving libpipeline1:amd64 (1.3.0-1) ...\nProcessing triggers for mime-support (3.54ubuntu1.1) ...\nProcessing triggers for libc-bin (2.19-0ubuntu6.9) ...\n",
    "stdout_lines": [
        "Reading package lists...",
        "Building dependency tree...",
        "Reading state information...",
        "The following packages will be REMOVED:",
        "  bsdmainutils groff-base libpipeline1 man-db",
        "0 upgraded, 0 newly installed, 4 to remove and 0 not upgraded.",
        "After this operation, 5635 kB disk space will be freed.",
        "(Reading database ... 22252 files and directories currently installed.)",
        "Removing man-db (2.6.7.1-1ubuntu1) ...",
        "Removing bsdmainutils (9.0.5ubuntu1) ...",
        "Removing groff-base (1.22.2-5) ...",
        "Removing libpipeline1:amd64 (1.3.0-1) ...",
        "Processing triggers for mime-support (3.54ubuntu1.1) ...",
        "Processing triggers for libc-bin (2.19-0ubuntu6.9) ..."
    ]
}
```

### CODE REVIEW

File: packaging/os/apt.py [hunk 1]
```diff
@@ -104,12 +104,12 @@
...
     description:
-      - If C(yes), remove unused dependency packages for all module states except I(build-dep).
+      - If C(yes), remove all unused dependency packages for all module states.
+      - If C(yes) and package(s) name are provided, remove all provided package(s) with unused dependency.
     required: false
     default: no
     choices: [ "yes", "no" ]
-    aliases: [ 'autoclean']
-    version_added: "2.1"
+    version_added: "2.3"
```
Comment (resmo) [inline packaging/os/apt.py]: do not change version_added as the param is available since 2.1

File: packaging/os/apt.py [hunk 2]
```diff
@@ -794,9 +845,9 @@ def main():
             force = dict(default='no', type='bool'),
             upgrade = dict(choices=['no', 'yes', 'safe', 'full', 'dist']),
             dpkg_options = dict(default=DPKG_OPTIONS),
-            autoremove = dict(type='bool', default=False, aliases=['autoclean']),
```
Comment (resmo) [inline packaging/os/apt.py:797]: autoclean alias should not be removed, this breaks backwards compatibly,

**Comment by gregdek** [comment]
Thanks @gdelpierre for this PR. This module is maintained by the Ansible core team, so it can take a while for patches to be reviewed. Thanks for your patience.

Core team: please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with 'needs_revision' or merge as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by bcoca** [comment]
fixed via https://github.com/ansible/ansible/pull/22154

**Pull request closed without merging**